### PR TITLE
Use py-cpuinfo on windows

### PR DESCRIPTION
**Pull request by isuruf**
Alternative to https://github.com/archspec/archspec/pull/114. This gives flags in addition to the other info in #114

### CODE REVIEW

**Comment by isuruf** [comment]
Can someone approve the workflow for github actions to be run?

**Comment by alalazo** [comment]
@isuruf We are going to check this one within the next two weeks. Feel free to ping me at the end of next week if everything stays silent.

File: archspec/cpu/detect.py [hunk 1]
```diff
@@ -77,20 +77,49 @@ def proc_cpuinfo():
     return info
 
 
+@info_dict(operating_system="Windows")
+def proc_py_cpuinfo():
+    """Returns a raw info dictionary by using py-cpuinfo"""
+    try:
+        import cpuinfo
+    except ImportError:
+        return {}
```
Comment (isuruf) [inline archspec/cpu/detect.py:86]: Linter is complaining that this import is not at the top-level. Should I explicitly ignore that warning?

**Comment by isuruf** [comment]
Ping on this @alalazo

**Comment by alalazo** [comment]
@johnwparent @scheibelp This is the PR adding host detection for Windows we talked briefly about. It uses https://github.com/workhorsy/py-cpuinfo which also has linux support. There's a prior attempt of adding Windows support in #114 too.

**Comment by johnwparent** [comment]
@isuruf Thanks for the contribution! Exciting to see more Windows users!
I'll be sure to ping you if I have any questions, I'll be digging into this over the coming week.

**Comment by isuruf** [comment]
@johnwparent, any progress on this?

**Comment by alalazo** [comment]
@johnwparent If we use `py-cpuinfo` we should probably use it consistently across `archspec`.

`py-cpuinfo` doesn't seem to have other Python dependencies itself, but it might rely on a few system tools for linux, if they are present. This can make detection non-deterministic (we have seen the same with the `distro` package we use to detect the OS - having or not a tool slightly changes the distro name).

I guess one thing we might want to check is how difficult it would be to code something in archspec to perform the same detection as `py-cpuinfo` does on Windows. Apparently the only method used there is `Windows Registry`.

**Comment by johnwparent** [comment]
@alalazo @scheibelp
> @johnwparent If we use `py-cpuinfo` we should probably use it consistently across `archspec`.
> 

Agreed
> `py-cpuinfo` doesn't seem to have other Python dependencies itself, but it might rely on a few system tools for linux, if they are present.
>

Very frankly, `py-cpuinfo` hasn't been updated for two years and the project seems dead. At the very least, it seems like community contributions are stagnating pretty quickly and we'd be unable to rely on updates. If we don't need updates that's one thing, but IMO it's a point against using it.


> This can make detection non-deterministic (we have seen the same with the `distro` package we use to detect the OS - having or not a tool slightly changes the distro name).
> 
we probably want to shy away from non-determinism (and personally I dislike relying on system packages to exist). Another consideration is whether or not this supports all the micro-archs we support, like `cpuinfo` (the c++ package) doesn't support powerpc, which IIRC, we need to support.

> I guess one thing we might want to check is how difficult it would be to code something in archspec to perform the same detection as `py-cpuinfo` does on Windows. Apparently the only method used there is `Windows Registry`.

Windows detection is likely going to be less complicated than unix based systems (shocking, right?) because Windows takes a pretty opinionated stance in terms of detection for its own tools that we can leverage. Should be relatively straightforward to implement on our own. I know CMake does a very good job of this on Windows so I'll see how we do it there.

tl;dr I think we should avoid using `py-cpuinfo` for a number of reasons and implement what needs to be implemented in archspec directly.

**Comment by alalazo** [comment]
> I think we should avoid using py-cpuinfo for a number of reasons and implement what needs to be implemented in archspec directly.

:handshake: I concur. For sure we want to avoid non-determinism. If it's easy to implement detection for Windows we should do it in `archspec`, and avoid an extra dependency.

**Comment by isuruf** [comment]
> Another consideration is whether or not this supports all the micro-archs we support, like cpuinfo (the c++ package) doesn't support 64 bit powerpc, which IIRC, we need to support.

The python package `py-cpuinfo` supports 64-bit powerpc and other architectures like loongson64.

**Comment by alalazo** [comment]
@isuruf I just tried to compare:
```console
$ time archspec cpu
icelake

real	0m0,101s
user	0m0,072s
sys	0m0,006s
```
and
```console
$ time python -m cpuinfo
Python Version: 3.8.10.final.0 (64 bit)
Cpuinfo Version: 9.0.0
Vendor ID Raw: GenuineIntel
Hardware Raw: 
Brand Raw: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
Hz Advertised Friendly: 3.0000 GHz
Hz Actual Friendly: 1.8000 GHz
Hz Advertised: (3000000000, 0)
Hz Actual: (1800000000, 0)
[ ... ]
real	0m1,181s
user	0m1,098s
sys	0m0,069s
```
One second is too much for us to use `cpuinfo` as is.[^1] I think we are better off coding what we need for Windows directly in `archspec`. Would you be willing to take a stab at it? It amounts to code something similar to:

https://github.com/archspec/archspec/blob/8e3b4dd24c4d53c83f09d4fc70cfeee2787df408/archspec/cpu/detect.py#L111-L148

but for Windows. We can try to work with #114, if the detection method there seems the right approach, or code something different. @johnwparent Do you have any idea already?

[^1]: Not to mention that the project seems stale, and hardly any issue gets a reply these days.

**Comment by isuruf** [comment]
#114 does not provide information about the micro-architecture which is what I need for conda.

>  think we are better off coding what we need for Windows directly in archspec. Would you be willing to take a stab at it? It amounts to code something similar to:

What do you think about vendoring `py-cpuinfo` as I have done here and work on improving it if needed?

**Comment by alalazo** [comment]
> What do you think about vendoring py-cpuinfo as I have done here and work on improving it if needed?

I'd rather check how to extract the information we need on Windows, and code it in `archspec` directly. Not really in favor of vendoring 2.8k lines of unmaintained code, when we need just a tiny fraction of it.

**Comment by johnwparent** [comment]
> #114 does not provide information about the micro-architecture which is what I need for conda.
> 

#114 Isn't the solution we'll be moving forward with, we'll either be using the registry or the win32api, as those are, AFAIK, the idiomatic methods to detect arch on Windows.

> What do you think about vendoring `py-cpuinfo` as I have done here and work on improving it if needed?

I won't speak for @alalazo but I don't see a ton of returned value on maintaining an entire python project that we do not have the history for simply to support Windows when there are other options.

**Comment by isuruf** [comment]
Okay. I'll extract the windows bits from py-cpuinfo.

**Comment by johnwparent** [comment]
@alalazo In the event that extracting the detection from `py-cpuinfo` retains the performance deficits you observed, what margin w.r.t. slowdown is acceptable for `archspec`?
There is a way to detect arch with the win32api (either the python module or by using `ctypes` and calling the win32api directly) which is likely a bit faster.

File: archspec/_vendor/cpuinfo/cpuinfo.py [hunk 2]
```diff
@@ -0,0 +1,2839 @@
...
+			retval = get_ticks_x86_64
+		return retval
+
+	def get_raw_hz(self):
+		from time import sleep
+
+		ticks_fn = self.get_ticks_func()
+
+		start = ticks_fn.func()
+		sleep(1)
```
Comment (alalazo) [inline archspec/_vendor/cpuinfo/cpuinfo.py:1526]: Here's the slowdown suspect n. 1 :laughing:

Comment (alalazo) [reply archspec/_vendor/cpuinfo/cpuinfo.py:1526]: > Okay. I'll extract the windows bits from py-cpuinfo.

Thanks @isuruf

Comment (johnwparent) [reply archspec/_vendor/cpuinfo/cpuinfo.py:1526]: > Here's the slowdown suspect n. 1 😆 (I still have to profile, but it seems likely we are hitting this code path)

🤣 🤣 🤣 Alright so it seems like we're safe in that regard.

**Comment by isuruf** [comment]
Extracted cpuid bits from py-cpuinfo. Please have a look

File: archspec/cpu/cpuid.py [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+
+    # Return {} if this is not an X86 CPU
+    if not _is_x86():
+        return {}
+
+    try:
+        # Start running the function in a subprocess
+        queue = Queue()
+        p = Process(target=_get_cpu_info_from_cpuid_subprocess_wrapper, args=(queue,))
+        p.start()
```
Comment (alalazo) [inline archspec/cpu/cpuid.py:676]: Why do we need to spawn a new process at all?  Can't we encode a much simpler function, detached from the `py-cpuinfo` implementation?

File: archspec/cpu/cpuid.py [hunk 4]
```diff
@@ -0,0 +1,697 @@
...
+        asm = ASM(restype, argtypes, machine_code)
+        asm.compile()
+        return asm
+
+    def _run_asm(self, *machine_code):
+        asm = ASM(ctypes.c_uint32, (), machine_code)
+        asm.compile()
+        retval = asm.run()
+        asm.free()
+        return retval
```
Comment (alalazo) [inline archspec/cpu/cpuid.py:207]: Do we really need to run assembly code? Isn't there any kind of simple inspection that we can do on Windows? @johnwparent

**Review by alalazo — CHANGES_REQUESTED** [review]
I think this is still somehow vendoring `py-cpuinfo`, which is unwanted. 

@johnwparent Can you check if there is any mechanism to detect what we need on Windows that doesn't need running ASM code?

File: archspec/cpu/cpuid.py [hunk 5]
```diff
@@ -0,0 +1,697 @@
...
+        processor_brand = processor_brand.strip("\0").strip()
+
+        return processor_brand
+
+
+def _get_cpu_info_from_cpuid_actual():
+    """
+    Warning! This function has the potential to crash the Python runtime.
+    Do not call it directly. Use the get_cpu_info_from_cpuid function instead.
+    It will safely call this function in another process.
```
Comment (isuruf) [inline archspec/cpu/cpuid.py:621]: @alalazo, the ASM function might crash the python runtime

File: archspec/cpu/cpuid.py (hunk 4 above)
Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: @isuruf is `wmic cpu list` not sufficient? What detail is lacking from that command? Additionally, there are other methods to query the registry, i.e. `winreg`, would that work?

Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: @isuruf to save you some time, this is what `wmic cpu list /format:list` gives me for my machine:
```
AddressWidth=64
Architecture=9
Availability=3
Caption=Intel64 Family 6 Model 158 Stepping 13
ConfigManagerErrorCode=
ConfigManagerUserConfig=
CpuStatus=1
CreationClassName=Win32_Processor
CurrentClockSpeed=2304
CurrentVoltage=10
DataWidth=64
Description=Intel64 Family 6 Model 158 Stepping 13
DeviceID=CPU0
ErrorCleared=
ErrorDescription=
ExtClock=100
Family=207
InstallDate=
L2CacheSize=2048
L2CacheSpeed=
LastErrorCode=
Level=6
LoadPercentage=8
Manufacturer=GenuineIntel
MaxClockSpeed=2304
Name=Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
OtherFamilyDescription=
PNPDeviceID=
PowerManagementCapabilities=
PowerManagementSupported=FALSE
ProcessorId=BFEBFBFF000906ED
ProcessorType=3
Revision=
Role=CPU
SocketDesignation=U3E1
Status=OK
StatusInfo=3
Stepping=
SystemCreationClassName=Win32_ComputerSystem
SystemName=DVORAK
UniqueId=
UpgradeMethod=52
Version=
VoltageCaps=
```

Comment (isuruf) [reply archspec/cpu/cpuid.py:207]: @johnwparent, processorid is the only information there which can be used for feature detection. It contains EAX, EDX values and model numbers. We need `EBX` and `ECX` values from cpuid for detection of things like avx2, avx512.
The model number is useful, but then we would need to carry a database of model numbers to features which we would have to update for each model that each hardware vendor produces and is not feasible.

Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: `wmic` suggestion aside, after more investigation I see no other way to get cpu flags, seems to be the only way is C preproc/C program/Assembly via `CPUID` which is how `/proc/cpuinfo` is created anyway.

Comment (alalazo) [reply archspec/cpu/cpuid.py:207]: @isuruf For your use case, do you need feature parity with `linux` on detection, or detecting x86_64 levels would suffice?

Comment (isuruf) [reply archspec/cpu/cpuid.py:207]: Yeah, the levels would suffice, but that API is limited and we can't distinguish Knight's landing vs x86_64 level 4 since there's no way to check for `avx512vl` and `avx512dq` if I'm reading the docs correctly.

Comment (alalazo) [reply archspec/cpu/cpuid.py:207]: This seems interesting: https://github.com/albertony/cpufeatures?tab=readme-ov-file#cpufeatures-1

Comment (alalazo) [reply archspec/cpu/cpuid.py:207]: This snippet is interesting too: https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=msvc-170

Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: There's also `!cpuinfo` which is the assembler instruction that provides a lot of this information.

Comment (isuruf) [reply archspec/cpu/cpuid.py:207]: > This seems interesting: https://github.com/albertony/cpufeatures?tab=readme-ov-file#cpufeatures-1

> This snippet is interesting too: https://learn.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=msvc-170

Both of these would require compiling a C file which would complicate things. This basically runs the same ASM as is done by py-cpuinfo.

> There's also !cpuinfo which is the assembler instruction that provides a lot of this information.

You mean `cpuid`? That's what `py-cpuinfo` and others use.

Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: Sorry, yes I did mean to write `cpuid`, but I did also mean to include `!cpuinfo`, which is an extension for the kernel debugger, just mashed the thoughts into one sentence.

Comment (johnwparent) [reply archspec/cpu/cpuid.py:207]: @alalazo, outside of the significant amount of `py-cpuinfo` code we're vendoring here (I think we both agree that we should roll our own approach here) is the concern with the compiler/assembler the fragility or the performance (or perhaps both)?

**Comment by isuruf** [comment]
I reduced down the complexity a lot and it's down to 540 lines out of which 220 is for all the flags and fair bit of lines are just comments/LICENSE info.

**Comment by alalazo** [comment]
Two questions (forgive the lack of windows knowledge):
1. Would it be fine if we require users to have `wsl` available?
2. If we have `wsl` can we retrieve this information more easily (i.e. without maintaining ourselves the machine code to test `cpuid` and see which features are enabled)?

@johnwparent 

> is the concern with the compiler/assembler the fragility or the performance (or perhaps both)?

My main concern is maintenance. With `linux` and `darwin` we have a detection based on information that are returned by the OS. This means that we rarely need to update the host detection code, and we are ensured that CPU features can be used by the OS.

Following the `cpuid` approach on Windows, we need to update the detection mechanism for each and every new feature we want to detect. We are, as far as I understand, limited to `x86_64` architectures and we query the CPU directly - so there's no feedback that the OS can actually use the feature[^1] 

I still hope we could find an easier solution for Windows...

[^1]: This might be a minor problem, and I'm not sure what to expect from Windows. We encountered a few times situations were very old rhel systems installed on recent cpus had to "downgrade" the microachitecture (because of lacking kernel support). This was exposed as missing features in `/proc/cpuinfo`.

**Comment by isuruf** [comment]
> Would it be fine if we require users to have wsl available?

No, that's not really an option for me. That's a much bigger dependency than using a python library.

**Comment by johnwparent** [comment]
> 1. Would it be fine if we require users to have `wsl` available?

Frankly, no. It would be the same as requiring them to have Docker, virtualbox, vmware, or podman with slightly fewer steps. Additionally, if we're running this on a machine running Windows server, WSL is not installed by default (which admittedly is insanely easy to do, I just don't think `archspec` should be doing that for users). 

> 2. If we have `wsl` can we retrieve this information more easily (i.e. without maintaining ourselves the machine code to test `cpuid` and see which features are enabled)?

You'd have to have `archspec` spin up and then access a running VM (or figure out how to get into it's filesystem on host disc when not running) which is a decidedly non trivial effort, and IMO order of magnitude more complex and fragile than what @isuruf is doing here. If your concern is maintenance, then this will be substantively harder to maintain than some ASM/C code.

> My main concern is maintenance. With `linux` and `darwin` we have a detection based on information that are returned by the OS. This means that we rarely need to update the host detection code, and we are ensured that CPU features can be used by the OS.
> 
> Following the `cpuid` approach on Windows, we need to update the detection mechanism for each and every new feature we want to detect. We are, as far as I understand, limited to `x86_64` architectures and we query the CPU directly - so there's no feedback that the OS can actually use the feature[1](#user-content-fn-1-2e90e376bde7adc0723d6e8328b3b95a)
> 
> I still hope we could find an easier solution for Windows...

@isuruf Is there no amalgamation of functions in the Win32API that would acquire the info you need? 
Similar to how CPUINFO is generated? https://elixir.bootlin.com/linux/v5.14.13/source/arch/x86/kernel/cpu/common.c#L641

That would at least address @alalazo's concern w.r.t. having the OS determine features rather than the CPU itself, and IMO, this approach would be easier to maintain.

@alalazo w.r.t. maintenance I think this approach (if feasible) would also reduce the need for frequent updates. Additionally, I'm happy to maintain the Windows side of this as needed since we're going to need better `archspec` detection on Windows for Spack anyway (also with the mingw cross compiler eventually coming to Spack, we'll really need this for Spack on Windows)

**Comment by isuruf** [comment]
> @isuruf Is there no amalgamation of functions in the Win32API that would acquire the info you need?

I looked a lot, and couldn't find any. The closest was https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-isprocessorfeaturepresent, but it lacks a few flags to differentiate x86_64_v3 from x86_64_v4.

**Comment by isuruf** [comment]
FWIW, I'll be able to help with maintenance as well.

**Comment by alalazo** [comment]
> but it lacks a few flags to differentiate x86_64_v3 from x86_64_v4.

What would be missing? Is it the Knights landing vs. Skylake issue you were mentioning here https://github.com/archspec/archspec/pull/129#discussion_r1463725068 or is there more?

**Comment by isuruf** [comment]
> What would be missing? Is it the Knights Landing vs. Skylake issue you were mentioning here https://github.com/archspec/archspec/pull/129#discussion_r1463725068 or is there more?

Yes, that is it.

**Comment by johnwparent** [comment]
@isuruf Would a mixed bag of CPU info and C API's be best then? If we're only detecting a few set features from the assembler that are arch detection (rather than a feature an OS kernel would need to decide to implement) I think that would keep the maintenance burden w.r.t. detection limited and still allow for us to get what you need. @alalazo thoughts?

@isuruf is it just the v3 vs v4 issue or is that just a good example indicating a larger issue?

**Comment by alalazo** [comment]
> Yes, that is it.

Can we use some heuristic to solve that? For instance, `AVX512F` is a feature and:
- number of cores > 64, or
- some "brand string" contains "Xeon Phi", or
- know model and stepping from [https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info](SystemInfo) that match Knights Landing

Do you see issues with an approach like that (I assume that Win32 API is wrapped in Python, right?)

**Comment by johnwparent** [comment]
> > Yes, that is it.
> 
> Can we use some heuristic to solve that? For instance, `AVX512F` is a feature and:
> 
> * number of cores > 64, or
> * some "brand string" contains "Xeon Phi", or
> * know model and stepping from [https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info](SystemInfo) that match Knights Landing

Isn't this effectively just maintaining a DB of associated properties? We could do that and just use WMIC, no need for C/ASM.

> Do you see issues with an approach like that (I assume that Win32 API is wrapped in Python, right?)

... Sort of, parts of it are wrapped but we can get access to all of it by just loading whatever libraries we need via the `ctypes` library

**Comment by alalazo** [comment]
> Isn't this effectively just maintaining a DB of associated properties? 

The idea is to store just what we need to detect the outlier (Xeon Phi, discontinued as of 2018), if we're lucky just a string. 

I'll try to summarize my understanding so far:
1. The current approach uses `cpuid`, bypasses the OS, but is rather complete in detecting features. It's only supported for x86_64 and we'll need to update the detection logic with new inspections as new CPUs come out.
2. Alternatively, we could use [this](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-isprocessorfeaturepresent), if we find a way to distinguish Xeon Phi from other CPUs with AVX512F support. In this way we'll only support generic architectures, but we'll likely have a simpler implementation. This approach seems to give minimal support for `aarch64` too.

Does that seem correct?

**Comment by isuruf** [comment]
For approach 2, you can get model, processor from wmic, but that bypasses the amd64 emulator on Win-ARM64 boxes too. On a Win-ARM64 box running x86_64 python, I get the values for aarch64 which is confusing. So, both options bypass the OS.

**Comment by isuruf** [comment]
With the branch https://github.com/isuruf/archspec/tree/windows2 which uses wmic and IsProcessorFeaturePresent, I get 
```
{'cpu family': '8', 'model': 'D4B', 'processor': '3', 'vendor_id': 'Qualcomm Technologies Inc', 'model name': 'Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz', 'flags': 'lahf_lm sse2 sse4_1 sse4_2 popcnt cx16 mmx sse ssse3'}
```
which is really confusing.

With this PR, I get on the same Win-ARM64 box
```
{'vendor_id': 'AuthenticAMD', 'model name': 'Virtual CPU @ 2.99GHz', 'model': 4, 'cpu family': 15, 'flags': '3dnowprefetch aes clflush cmov cx16 cx8 fpu fxsr ht lahf_lm mmx pclmulqdq pni popcnt sse sse2 sse4_1 sse4_2 ssse3 tsc'}
```

**Comment by alalazo** [comment]
The case of an emulated x86_64 on an ARM box is similar to #77  - so we may add an option to detect the emulated architecture. In any case, from the branch, it seems you are adding the flags - so I would say it's a bug to add them based on `platform.machine()` when the CPU is retrieved with `wmic`.

File: archspec/cpu/cpuid.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        0x89, 0x07,              # mov    %eax,(%edi)
+        0x89, 0x5f, 0x04,        # mov    %ebx,0x4(%edi)
+        0x89, 0x4f, 0x08,        # mov    %ecx,0x8(%edi)
+        0x89, 0x57, 0x0c,        # mov    %edx,0xc(%edi)
+        0x5f,                    # pop    %edi
+        0x5b,                    # pop    %ebx
+        0xc3                     # ret
+]
+
+is_windows = os.name == "nt"
```
Comment (alalazo) [inline archspec/cpu/cpuid.py:90]: Can you explain this a bit more broadly? Which values can we expect from `os.name`? Also, why aren't we just checking `platform.system()`?

Comment (alalazo) [reply archspec/cpu/cpuid.py:90]: Ok, I realize now this is taken from https://github.com/flababah/cpuid.py I guess we don't really need all the non-windows part.

File: archspec/cpu/cpuid.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+is_64bit   = ctypes.sizeof(ctypes.c_voidp) == 8
+
+class CPUID_struct(ctypes.Structure):
+    _fields_ = [(r, c_uint32) for r in ("eax", "ebx", "ecx", "edx")]
+
+class CPUID(object):
+    def __init__(self):
+        if platform.machine() not in ("AMD64", "x86_64", "x86", "i686"):
+            raise SystemError("Only available for x86")
```
Comment (alalazo) [inline archspec/cpu/cpuid.py:99]: Do you have any idea how to deal with `aarch64`?

**Review by alalazo — COMMENTED** [review]
I have a few questions below. 

If we proceed with this approach, I think we should design it in a way that the bit and registers to be checked are read from a JSON file. We'll keep in Python only the code injection to run ASM code. 

Does that seem reasonable?

Also, let me know if you are fine with us pushing code here, or if you prefer to us to experiment in another PR / branch.

File: archspec/cpu/cpuid.py (hunk 7 above)
Comment (isuruf) [reply archspec/cpu/cpuid.py:99]: Not at the moment, but aarch64 seems to have a cpuid flag that we can possibly use.

**Comment by isuruf** [comment]
> Does that seem reasonable?

Yep. Sounds good. Pushed a change for this.

> Also, let me know if you are fine with us pushing code here, or if you prefer to us to experiment in another PR / branch.

Sure

File: archspec/cpu/cpuid_flags.json [hunk 8]
```diff
@@ -0,0 +1,622 @@
+{
+  "fpu": {
+    "eax": 1,
+    "edx": 0
+  },
```
Comment (alalazo) [inline archspec/cpu/cpuid_flags.json:5]: I changed the format of the JSON in #150 . This one has duplicate keys (for `eax=0x1` and `eax=0x80000001`). I also moved other constants we need in the JSON file.

File: archspec/cpu/cpuid_flags.json [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+    "ecx": 22
+  },
+  "sgx_lc": {
+    "eax": 7,
+    "ecx": 30
+  },
+  "fpu": {
+    "eax": 2147483649,
+    "ebx": 0
+  },
```
Comment (alalazo) [inline archspec/cpu/cpuid_flags.json:413]: Looking at https://en.wikipedia.org/wiki/CPUID#EAX=80000001h:_Extended_Processor_Info_and_Feature_Bits I think this and the items below are wrong. The bug comes directly from `cpuinfo`, where `ebx` should really be `edx`.
